File: app/src/main/java/com/weichen2046/filesender2/service/ServiceManager.java
package com.weichen2046.filesender2.service;

import android.app.Service;
import android.content.Intent;
import android.os.IBinder;
import android.os.RemoteException;
import android.util.Log;
import android.util.SparseArray;

import com.weichen2046.filesender2.utils.TcpDataSender;

public class ServiceManager extends Service {

    private static final String TAG = "ServiceManager";

    private SparseArray<IBinder> mSubServices = new SparseArray<>();

    public static final int SERVICE_DESKTOP_DISCOVERER  = 1;
    public static final int SERVICE_UDP_DATA_MONITOR    = 2;
    public static final int SERVICE_TCP_DATA_MONITOR    = 3;
    public static final int SERVICE_DESKTOP_MANAGER     = 4;

    public ServiceManager() {
    }

    @Override
    public IBinder onBind(Intent intent) {
        return mBinder;
    }

    private IBinder mBinder = new IServiceManager.Stub() {
        @Override
        public synchronized IBinder getService(int serviceId) throws RemoteException {
            IBinder binder = mSubServices.get(serviceId);
            if (binder == null) {
                if (serviceId == SERVICE_DESKTOP_DISCOVERER) {
                    binder = new DesktopDiscoverer();
                    mSubServices.put(SERVICE_DESKTOP_DISCOVERER, binder);
                }

                if (serviceId == SERVICE_UDP_DATA_MONITOR) {
                    UdpDataMonitor udpDataMonitor = new UdpDataMonitor();
                    udpDataMonitor.attach(IServiceManager.Stub.asInterface(this));
                    binder = udpDataMonitor;
                    mSubServices.put(SERVICE_UDP_DATA_MONITOR, binder);
                }

                if (serviceId == SERVICE_TCP_DATA_MONITOR) {
                    TcpDataMonitor tcpDataMonitor = new TcpDataMonitor();
                    tcpDataMonitor.attach(IServiceManager.Stub.asInterface(this));
                    binder = tcpDataMonitor;
                    mSubServices.put(SERVICE_TCP_DATA_MONITOR, binder);
                }

                if (serviceId == SERVICE_DESKTOP_MANAGER) {
                    binder = new DesktopManager();
                    mSubServices.put(SERVICE_DESKTOP_MANAGER, binder);
                }
            }
            return binder;
        }
    };

    @Override
    public void onDestroy() {
        Log.d(TAG, "onDestroy called.");
        super.onDestroy();
    }

    @Override
    public boolean onUnbind(Intent intent) {
        Log.d(TAG, "onUnbind called.");
        return super.onUnbind(intent);
    }
}

File: app/src/main/java/com/weichen2046/filesender2/service/ServiceManager.kt
package com.weichen2046.filesender2.service

import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.os.IBinder

/**
 * Created by chenwei on 2017/6/12.
 */
object ServiceManager {
    var serviceManager: IServiceManager? = null
        private set

    var remoteDevicesManager: RemoteDevicesManager? = null
        private set

    var udpDataMonitor: IUdpDataMonitor? = null
        private set

    var tcpDataMonitor: ITcpDataMonitor? = null
        private set

    var deviceDiscoverer: IRemoteDeviceDiscoverer? = null
        private set

    private var connected: Boolean = false

    fun init(context: Context, connectedCallback: () -> Unit, disconnectedCallback: () -> Unit) {
        if (connected) {
            connectedCallback()
            return
        }
        val serviceIntent = Intent(context, ServiceManagerInternal::class.java)
        context.bindService(serviceIntent, object: ServiceConnection {
            override fun onServiceDisconnected(name: ComponentName?) {
                connected = false
                serviceManager = null
                remoteDevicesManager = null
                udpDataMonitor = null
                tcpDataMonitor = null
                deviceDiscoverer = null
                disconnectedCallback()
            }

            override fun onServiceConnected(name: ComponentName?, service: IBinder?) {
                connected = true
                serviceManager = IServiceManager.Stub.asInterface(service)
                var binder = serviceManager?.getService(ServiceManagerInternal.SERVICE_DEVICES_MANAGER)
                remoteDevicesManager = RemoteDevicesManager(IRemoteDevicesManager.Stub.asInterface(binder))
                binder = serviceManager?.getService((ServiceManagerInternal.SERVICE_UDP_DATA_MONITOR))
                udpDataMonitor = IUdpDataMonitor.Stub.asInterface(binder)
                binder = serviceManager?.getService((ServiceManagerInternal.SERVICE_TCP_DATA_MONITOR))
                tcpDataMonitor = ITcpDataMonitor.Stub.asInterface(binder)
                binder = serviceManager?.getService((ServiceManagerInternal.SERVICE_DEVICE_DISCOVERER))
                deviceDiscoverer = IRemoteDeviceDiscoverer.Stub.asInterface(binder)
                connectedCallback()
            }
        }, Context.BIND_AUTO_CREATE)
    }
}